### PR TITLE
Document public metric SDK interfaces to remain stable

### LINKED ISSUE #3673: Document public interfaces with versioning policy
Similar to https://github.com/open-telemetry/opentelemetry-go/pull/2012, add documentation to exported interfaces.

### PR DESCRIPTION
**Pull request by MrAlias**
Resolve #3673

Similar to https://github.com/open-telemetry/opentelemetry-go/pull/2012, document for developers so they are aware things should not change in the SDK.

### CODE REVIEW

File: sdk/metric/aggregation/aggregation.go [hunk 1]
```diff
@@ -32,9 +32,13 @@ type Aggregation interface {
 
 	// Copy returns a deep copy of the Aggregation.
 	Copy() Aggregation
+	// DO NOT CHANGE: any modification will not be backwards compatible and
```
Comment (pellared) [inline sdk/metric/aggregation/aggregation.go:35]: We do not allow user-defined aggregations. Therefore should we add a following notice?

> Warning: methods may be added to this interface in minor releases.

File: sdk/metric/exporter.go [hunk 2]
```diff
@@ -34,12 +34,16 @@ type Exporter interface {
 	// This method needs to be concurrent safe with itself and all the other
 	// Exporter methods.
 	Temporality(InstrumentKind) metricdata.Temporality
+	// DO NOT CHANGE: any modification will not be backwards compatible and
+	// must never be done outside of a new major release.
```
Comment (pellared) [inline sdk/metric/exporter.go:38]: General comment

What is the value of these comments? Is it not obvious that (by default) an interface cannot change once the package is stable? Moreover we have https://github.com/open-telemetry/opentelemetry-go/blob/main/CONTRIBUTING.md#interface-stability

File: sdk/metric/reader.go [hunk 3]
```diff
@@ -56,13 +56,17 @@ type Reader interface {
 	// The producer argument allows the Reader to signal the sdk to collect
 	// and send aggregated metric measurements.
 	register(sdkProducer)
+	// DO NOT CHANGE: any modification will not be backwards compatible and
```
Comment (pellared) [inline sdk/metric/reader.go:59]: We do not allow user-defined readers. Therefore should we add a following notice?

> Warning: methods may be added to this interface in minor releases.

File: sdk/metric/aggregation/aggregation.go (hunk 1 above)
Comment (MrAlias) [reply sdk/metric/aggregation/aggregation.go:35]: This is an exported interface with exported methods. Modifying the exported methods in any way is a breaking change once they are released stability regardless of if we allow user defined aggregations or not. This comment is the conclusion of the community decision in #1714 to ensure we make it clear to future developers to not introduce breaking changes here.

File: sdk/metric/reader.go (hunk 3 above)
Comment (MrAlias) [reply sdk/metric/reader.go:59]: This is an exported interface with exported methods. Modifying the exported methods in any way is a breaking change once they are released stability regardless of if we allow user defined readers or not. This comment is the conclusion of the community decision in #1714 to ensure we make it clear to future developers to not introduce breaking changes here.

File: sdk/metric/exporter.go (hunk 2 above)
Comment (MrAlias) [reply sdk/metric/exporter.go:38]: This interface is not in the API and is not required to have methods added to it in the future. We therefore plan to address any change here in a backwards compatible Go manner.

**Comment by CharlieTLe** [comment]
Would it make sense to have tests that will fail if there are any new, modified, or removed exported methods?

File: sdk/metric/aggregation/aggregation.go (hunk 1 above)
Comment (pellared) [reply sdk/metric/aggregation/aggregation.go:35]: My comment was NOT about this line in particular but about adding a comment referenced below

https://github.com/open-telemetry/opentelemetry-go/blob/2899fcfdcaf0d5499d2659b93cee7e609da8c65d/CONTRIBUTING.md#L473-L476

As far as I understand in this case we could add a method to this interface in future as I think it falls into "Internal Only Interface Implementations".

File: sdk/metric/exporter.go (hunk 2 above)
Comment (pellared) [reply sdk/metric/exporter.go:38]: >  We therefore plan to address any change here in a backwards compatible Go manner.

That is why I would suggesting only adding comments to interfaces which are NOT handled in a backwards compatible Go manner. For me this comment only adds confusion 😉

Comment (MrAlias) [reply sdk/metric/exporter.go:38]: I'm not sure how adding a comment that says to not change things makes it unclear that that interface should not be changed. Can you explain?

Comment (pellared) [reply sdk/metric/exporter.go:38]: After the Go module is stable none method must be changed.

This comment makes me think: "Should this interface be handled differently?"

**Review by hanyuancheung — APPROVED** [review]

File: sdk/metric/reader.go (hunk 3 above)
Comment (pellared) [reply sdk/metric/reader.go:59]: Take notice that this is an interface based on specification so a method can be added.

**Comment by MrAlias** [comment]
From IRL conversation with @pellared :

- The `Aggregation` interface: is it neede?
  - Can we remove it and just export the concrete types?
  - The SDK can define its own private version of this interface to reference all the concrete types
  - The `Stream` will need a way to reference these types. It likely makes sense to change the field to `inteface{}` type
- The `Reader` interface may need to be extended based on future changes to the specification.
  - We probably don't want to document the interface cannot change, but instead need a plan for how we will extend it in the future.

File: sdk/metric/aggregation/aggregation.go (hunk 1 above)
Comment (pellared) [reply sdk/metric/aggregation/aggregation.go:35]: As far as I remember we should not allow user defined aggreagations. Right now, I think it is possible to hack (type embedding + different implementation of exported methods) to provide custom ones.

Maybe we should move the aggregation to `sdk/metrics` package and unexport all methods to not make it possible (similar approach to option interface).

We could also consider adding "type constraints" to only allow types defined by the SDK.

Comment (MrAlias) [reply sdk/metric/aggregation/aggregation.go]: Resolved: https://github.com/open-telemetry/opentelemetry-go/issues/4432

File: sdk/metric/exporter.go (hunk 2 above)
Comment (MrAlias) [reply sdk/metric/exporter.go:37]: This was a strategy the community decided on in https://github.com/open-telemetry/opentelemetry-go/issues/1714. This is ensuring consistency across signals and following the already instituted policy.

File: sdk/metric/reader.go (hunk 3 above)
Comment (MrAlias) [reply sdk/metric/reader.go]: Changed.

**Review by MadVikingGod — APPROVED** [review]

**Review by dmathieu — APPROVED** [review]

File: sdk/metric/reader.go [hunk 4]
```diff
@@ -50,6 +50,8 @@ var errNonPositiveDuration = fmt.Errorf("non-positive duration")
 //
 // Pull-based exporters will typically implement Register
 // themselves, since they read on demand.
+//
+// Warning: methods may be added to this interface in minor releases.
```
Comment (pellared) [inline sdk/metric/reader.go:54]: Shouldn't we add 

> // DO NOT CHANGE: any modification will not be backwards compatible and
	// must never be done outside of a new major release.

To the `Shutdown` method?

We can still add new methods but the existing should not be changed nor removed, right?

Comment (MrAlias) [reply sdk/metric/reader.go:54]: sure, sounds good

**Pull request merged**